### ISSUE TITLE
Word-based occurrence highlighting and suggestion in some writing systems are broken

### ISSUE THREAD

**Issue by Yarakashi-Kikohshi** [yzhang-gh/vscode-markdown#1092]
This problem does not occur when extensions are disabled, but it does occur when they are enabled.

# Occurrence highlighting and suggestions do not work well when `．` is present

`．`(`U+FF0E`) is not recognized as a punctuation mark and suggestions to the left of this character and occurrence highlighting do not work well. These features work well with `,` (`U+FF0C`).

In Japanese, writing `．` at the end of the sentence indicates the end of the sentence, like a period. `，` is used like a comma. Especially, it is inconvenient that the suggestion is not displayed after this punctuation because the text is typed after `．`.

As an example, I gave some demonstrations to the following sentence.

```
こんにちは．先生
```

## Demo 1: Occurrence highlighting

The entire text will be shaded. Should be separated by `．`.

![Place the cursor](https://i.gyazo.com/eba149db44c5bef8196562d4c31116dc.gif)

## Demo 2: Suggestion

Continue to type lestters and call suggestions (`Ctrl+Space`). In advance, I created `先生` and `teacher` as a snippet.

```json
{
  "Teacher in Japanese": {
    "prefix": "せんせい",
    "body": "先生",
    "description": "先生: Teacher in Japanese"
  },
  "Teacher in English": {
    "prefix": "teacher",
    "body": "Teacher",
    "description": "Teacher in English"
  }
}
```

Suggestions are not displayed. Even if you press `Ctrl+Space`, `No suggestions` are displayed.

![suggestions](https://i.gyazo.com/a2502c4215f5e239a110bc9aa9944e34.gif)

## Possible causes

I submitted this issue once to VS Code repository ([\# 144913][144913]). Apparently, this kind of problem may depend on `wordPattern`.

[144913]: https://github.com/microsoft/vscode/issues/144913

> https://github.com/microsoft/vscode/issues/144913#issuecomment-1066468990

## Version

- Markdown All in One: 3.4.0
- VS Code

    ```
     Version: 1.65.2 (user setup)
      Commit: c722ca6c7eed3d7987c0d5c3df5c45f6b15e77d1
        Date: 2022-03-10T14:33:55.248Z
    Electron: 13.5.2
    Chromium: 91.0.4472.164
     Node.js: 14.16.0
          V8: 9.1.269.39-electron.0
          OS: Windows_NT x64 10.0.19043
    ```

**Comment by Lemmingh** [comment]
Quite awkward.

We hard-code a small set of Punctuation to implement "quick styling":

<https://github.com/yzhang-gh/vscode-markdown/blob/c50fe848fae507597c2b1252c4dbe0708f16bd24/src/extension.ts#L62-L65>

**Comment by Lemmingh** [comment]
Unicode 14 defines more than 700 characters under Punctuation. I'm afraid it's impractical to inspect them one by one in each script.

**Comment by yzhang-gh** [comment]
I haven't looked at it but was thinking that we just take what we need. Not sure whether the only problem is on the `*`, or we may need to change how we "smart select" a word when there is no selection.

**Comment by Lemmingh** [comment]
Right. Perhaps we can use "inclusion":

"Normal word" plus `*` and `_` at proper position

---

A more robust (but currently not possible) solution may be:

Don't touch the word pattern; use parsing result to locate emphasis spans.

**Comment by yzhang-gh** [comment]
> Right. Perhaps we can use "inclusion":
> 
> "Normal word" plus `*` and `_` at proper position

I tried
```
([*_]{1,2})?(\p{Alphabetic}|\p{Number}|\p{Nonspacing_Mark})(((\p{Alphabetic}|\p{Number}|\p{Nonspacing_Mark})|[_])?(\p{Alphabetic}|\p{Number}|\p{Nonspacing_Mark}))*\1
```

![image](https://user-images.githubusercontent.com/7588612/158402714-fce77552-66ec-4fe3-bf90-2cad7bab2faa.png)

and it works well (between `こんにちは．先生|` and `こんにちは．**先生**|`)

---

The word suggestion seems to also work as expected. However, the occurrence highlighting (and double-click selection) does NOT follow our pattern. There is no highlight at all and double click the text will select `こんにちは．先生` as a whole. I suspect this may be an issue of VS Code.

---

> A more robust (but currently not possible) solution may be:
> 
> Don't touch the word pattern; use parsing result to locate emphasis spans.

True, but might not be worth it 😂

**Comment by yzhang-gh** [comment]
released in v3.4.2
